### PR TITLE
Fix race in Restart during Close task

### PR DESCRIPTION
**Pull request by BrennanConroy**

### CODE REVIEW

**Comment by BrennanConroy** [comment]
Needs a little bit of work still, the race just moved/got smaller

**Comment by anurse** [comment]
Hang on, doesn't this basically solve the race? We have start wait for a previous stop to complete before resetting the Close TCS.

**Comment by BrennanConroy** [comment]
Stop isn't the only thing that can cause a reconnect :(

File: src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs [hunk 1]
```diff
@@ -102,9 +108,239 @@ public HttpConnection(Uri url, ITransportFactory transportFactory, ILoggerFactor
...
+                        {
+                            _logger.ExceptionThrownFromCallback(nameof(OnReceived), ex);
+                        }
+                    }
+                };
+                _connection.Closed += ex =>
+                {
+                    ChangeState(from: ConnectionState.Connected, to: ConnectionState.Disconnected);
+                    var closed = _closed.ToArray();
+                    foreach (var callback in closed)
```
Comment (mikaelm12) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: New line

File: src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs [hunk 2]
```diff
@@ -27,32 +27,38 @@ public class HttpConnection : IConnection
...
+
+        private HttpConnectionOneOff _connection;
+        private List<Action<Exception>> _closed = new List<Action<Exception>>();
 
         public Uri Url { get; }
 
         public IFeatureCollection Features { get; } = new FeatureCollection();
 
-        public event Action<Exception> Closed;
+        public event Action<Exception> Closed
```
Comment (mikaelm12) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: Can we not just change _closed to be a Thread Safe type?

File: src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs [hunk 3]
```diff
@@ -102,9 +108,239 @@ public HttpConnection(Uri url, ITransportFactory transportFactory, ILoggerFactor
...
+        public async Task StopAsync() => await StopAsyncCore().ForceAsync();
+
+        private async Task StopAsyncCore()
+        {
+            HttpConnectionOneOff connection;
+            lock (_stateChangeLock)
+            {
+                if (!(_connectionState == ConnectionState.Connecting || _connectionState == ConnectionState.Connected))
+                {
+                    //_logger.SkippingStop(_connectionId);
```
Comment (mikaelm12) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: A bunch of commented out logging lines like this

File: src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs [hunk 4]
```diff
@@ -102,9 +108,239 @@ public HttpConnection(Uri url, ITransportFactory transportFactory, ILoggerFactor
...
+                }
+
+                _connectionState = ConnectionState.Disconnected;
+                connection = _connection;
+            }
+
+            await connection.StopAsync();
+        }
+
+        private ConnectionState ChangeState(ConnectionState from, ConnectionState to)
```
Comment (mikaelm12) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: Consider changing from - to to something like startState, transitionState

File: src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs [hunk 5]
```diff
@@ -102,9 +108,239 @@ public HttpConnection(Uri url, ITransportFactory transportFactory, ILoggerFactor
...
+                {
+                    ChangeState(from: ConnectionState.Connected, to: ConnectionState.Disconnected);
+                    var closed = _closed.ToArray();
+                    foreach (var callback in closed)
+                    {
+                        try
+                        {
+                            callback(ex);
+                        }
+                        catch (Exception)
```
Comment (mikaelm12) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: nit: Make this one line

File: src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs [hunk 6]
```diff
@@ -102,9 +108,239 @@ public HttpConnection(Uri url, ITransportFactory transportFactory, ILoggerFactor
...
+                {
+                    _connectionState = to;
+                }
+
+                //_logger.ConnectionStateChanged(_connectionId, state, to);
+                return state;
+            }
+        }
+
+        private ConnectionState ChangeState(ConnectionState to)
```
Comment (mikaelm12) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: Change param name

File: src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs (hunk 2 above)
Comment (BrennanConroy) [reply src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: Accessing shouldn't be frequent, in fact only 1 access should happen at a time in our code.

File: src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs [hunk 7]
```diff
@@ -102,9 +108,239 @@ public HttpConnection(Uri url, ITransportFactory transportFactory, ILoggerFactor
...
+        internal enum ConnectionState
+        {
+            Disconnected,
+            Connecting,
+            Connected,
+            Disposed
+        }
+    }
+
+    internal class HttpConnectionOneOff : IConnection
```
[the remark this answers is not in the document]
Comment (BrennanConroy) [reply src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: `HttpConnectionCore`
`HttpConnectionSession`

[the remark this answers is not in the document]
Comment (anurse) [reply src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: Let's go with `Session`.

File: src/Microsoft.AspNetCore.Sockets.Client.Http/Internal/HttpConnectionSession.cs [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+                            _closedEvent.Invoke(abortException);
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        // Suppress (but log) the exception, this is user code
+                        _logger.ErrorDuringClosedEvent(ex);
+                    }
+
+                    _closeTcs.SetResult(null);
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/Internal/HttpConnectionSession.cs]: FYI, I moved this to be after `_closedEvent` is called, this means closed callbacks will finish before `StopAsync` returns

**Comment by BrennanConroy** [comment]
New attempt

**Comment by BrennanConroy** [comment]
Ping
Review please

**Comment by anurse** [comment]
Still a bit nervous about the race that occurs when the transport closes itself. If the transport closes (triggering `Input.Completion`) and the user calls `StopAsync` there's a race. If the `Completion` callback runs to the end before `StopAsync` captures the `closeTcs`, it gets the `closeTcs` for the previous call.

We need to talk about this today with @davidfowl. I want to get a handle on these races and the behavior we're actually looking for.

**Comment by BrennanConroy** [comment]
Yeah we should definitely talk, but I'm pretty sure I fixed that race...

**Comment by anurse** [comment]
No, it can still happen any time between the start of `StopAsync` and where you capture the `closeTcs`

**Comment by BrennanConroy** [comment]
Thats why I added the check inside the lock and return

**Comment by anurse** [comment]
Ah, ok, we might be good there. Let's walk through it though.

**Pull request merged**